[Tests] Mutualize fixtures code in BaseHttpFixture

tlrx opened the pull request:
Many fixtures have similar code for writing the pid & ports files or for handling HTTP requests. This commit changes the current ExampleFixture in order to make it an abstract BaseHttpFixture class that can be extended for specific testing purposes. This class provides some simple Request and Response classes to represent the HTTP requests and responses and takes care of writing the pid and ports files and starting the HTTP server. It allows to remove duplicated code and forbidden APIs invocations.

The current AmazonS3Fixture, GoogleCloudStorageFixture, AzureStorageFixture, and URLFixture have been removed. The AmazonS3TestServer, AzureStorageTestServer, GoogleCloudStorageTestServer have been renamed and now extend BaseHttpFixture.

This commit also fixes an issue with the AntFixture that was expecting a false result for the waitCondition(), whereas it should expect `true`.

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/AntFixture.groovy:
```diff
@@ -149,11 +149,10 @@ public class AntFixture extends AntTask implements Fixture {
         boolean success
         try {
-            success = waitCondition(this, ant) == false
+            success = waitCondition(this, ant)
```
lol, good catch. Can you create a separate PR for this?

tlrx replied:
Sure, I'll do that.

tlrx replied:
I created https://github.com/elastic/elasticsearch/pull/31272

ywelsch commented on modules/reindex/build.gradle:
```diff
@@ -121,6 +121,10 @@ if (Os.isFamily(Os.FAMILY_WINDOWS)) {
            unzip.temporaryDir,
            version == '090'
+      waitCondition = { fixture, ant ->
+        return fixture.portsFile.exists()
```
why is this now needed?

tlrx replied:
The OldElasticsearch fixture used in the reindex module tests starts an Elasticsearch instance and waits for its HTTP service to be started, then it writes the HTTP port to the `ports` file.

Before this change, the fixture here used the default AntFixture's waitCondition which is to execute a HTTP GET request on the address indicated in the fixture's `ports` file. Most fixtures write a IP address and a port number (ex: 127.0.0.1:48356) in this file but OldElasticsearch only writes the Elasticsearch HTTP port. Without this change, the fixture executes a `GET http://9200` request and fails directly, which was unnoticed because the bug in the `waitCondition(this, ant) == false` evaluation.

I'll revert this change in this PR and address all changes related to  `waitCondition(this, ant) == false`  in another PR as your suggested.

ywelsch commented on test/fixtures/basic-fixture/src/main/java/basic/BaseHttpFixture.java:
```diff
@@ -0,0 +1,312 @@
+/**
+ * Base class for test fixtures that requires a {@link HttpServer} to work.
+ */
+public abstract class BaseHttpFixture {
```
should we put this into `test:framework` instead?

tlrx replied:
Yes we can.

tlrx replied:
But that would mean to bring back all the test framework classes and dependencies when we just want to start a single class to simulate an external service. Thinking about it twice, I don't think that's a good idea.

ywelsch replied:
sure, but looking at your changes here: https://github.com/elastic/elasticsearch/pull/31210/files#diff-59ba8e13c56cbfc8a1bad12a2c8df5b0R32
it seems that you've done exactly that, i.e., mix the fixture classpath into the test classpath.

Also, if we now have a separate project (base-fixture), we'll need to publish that dependency as well so that others can use it.

tlrx replied:
> it seems that you've done exactly that, i.e., mix the fixture classpath into the test classpath.

Yes, that is more a consequence of the fixture originally using a test server that was shared with other test class, and now with dedicated projects I could have move the fixture code to the main sources insead of test sources.

Anyway, moving the BasicFixture to the test framework simplifies things and, as you said, it also allows others to use it. So I did that.

tlrx commented:
@ywelsch Thanks for the review. I moved BasicFixture to the test framework and I renamed it AbstractHttpFixture. It simplifies things. I also merged master to get a chance to have CI passed. Can you have another look please?

tlrx commented:
Thanks @ywelsch
